src/search/search_engines/hierarchical_search_engines/iterative_width_search.cc

```diff
@@ -0,0 +1,144 @@
+#include "iterative_width_search.h"
+
+#include "goal_test.h"
+
+#include "../../option_parser.h"
+#include "../../plugin.h"
+#include "../../tasks/root_task.h"
+#include "../../tasks/propositional_task.h"
+#include "../../task_utils/successor_generator.h"
+#include "../../task_utils/task_properties.h"
+#include "../../utils/logging.h"
+#include "../../utils/memory.h"
+
+#include <cassert>
+#include <cstdlib>
+
+using namespace std;
+using namespace hierarchical_search_engine;
+
+namespace iw_search {
+IWSearch::IWSearch(const Options &opts)
+    : HierarchicalSearchEngine(opts),
+      width(opts.get<int>("width")),
+      debug(opts.get<utils::Verbosity>("verbosity") == utils::Verbosity::DEBUG),
+      m_novelty_base(nullptr),
+      m_novelty_table(0) {
+}
+
+bool IWSearch::is_novel(const State &state) {
+    return m_novelty_table.insert(dlplan::novelty::TupleIndexGenerator(m_novelty_base, m_propositional_task->get_fact_ids(state)), true);
+}
+
+bool IWSearch::is_novel(const OperatorProxy &op, const State &succ_state) {
+    return m_novelty_table.insert(dlplan::novelty::TupleIndexGenerator(m_novelty_base, m_propositional_task->get_fact_ids(op, succ_state)), true);
+}
+
+void IWSearch::print_statistics() const {
+    statistics.print_detailed_statistics();
+    m_search_space->print_statistics();
+}
+
+SearchStatus IWSearch::step() {
+    if (open_list.empty()) {
+        utils::g_log << "Completely explored state space -- no solution!" << endl;
+        return FAILED;
+    }
+    StateID id = open_list.front();
+    open_list.pop_front();
+    State state = m_state_registry->lookup_state(id);
+    // std::cout << "Initial state: " << m_propositional_task->compute_dlplan_state(state).str() << std::endl;
+    SearchNode node = m_search_space->get_node(state);
+    node.close();
+    assert(!node.is_dead_end());
+    statistics.inc_expanded();
+
+    /* Goal check in initial state. */
+    if (id == m_initial_state_id) {
+        if (m_goal_test->is_goal(m_state_registry->lookup_state(m_initial_state_id), state)) {
```

Comment:
Can be simplified to `if (id == m_initial_state_id && m_goal_test->is_goal(state, state))`.